payments/models.py

```diff
@@ -154,3 +169,348 @@ def __str__(self):
+    def berth_orders(self):
+        return self.get_queryset().filter(
+            Q(
+                Q(_product_content_type__model=BerthProduct._meta.model_name)
+                | Q(_lease_content_type__model=BerthLease._meta.model_name)
+            )
+        )
+
+    def winter_storage_orders(self):
+        return self.get_queryset().filter(
+            Q(
+                Q(_product_content_type__model=WinterStorageProduct._meta.model_name)
+                | Q(_lease_content_type__model=WinterStorageLease._meta.model_name)
+            )
+        )
+
+
+class Order(UUIDModel, TimeStampedModel):
+    customer = models.ForeignKey(
+        "customers.CustomerProfile",
+        verbose_name=_("customer"),
+        on_delete=models.CASCADE,
+        related_name="orders",
+    )
+    product = GenericForeignKey("_product_content_type", "_product_object_id")
+    lease = GenericForeignKey("_lease_content_type", "_lease_object_id")
+    status = EnumField(OrderStatus, default=OrderStatus.WAITING)
+    comment = models.TextField(blank=True, null=True)
+    price = models.DecimalField(
+        verbose_name=_("price"),
+        max_digits=10,
+        decimal_places=2,
+        validators=[MinValueValidator(Decimal("0.01"))],
+        blank=True,
+    )
+    tax_percentage = models.DecimalField(
+        verbose_name=_("tax percentage"), max_digits=5, decimal_places=2, blank=True,
+    )
+    due_date = models.DateField(
+        verbose_name=_("due date"), default=calculate_order_due_date,
+    )
+
+    _product_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="product",
+    )
+    _product_object_id = models.UUIDField(
+        null=True, blank=True, verbose_name=_("product")
+    )
+
+    _lease_content_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+        related_name="lease",
+    )
+    _lease_object_id = models.UUIDField(null=True, blank=True, verbose_name=_("lease"))
+
+    objects = OrderManager()
+
+    def __str__(self):
+        return f"{self.product} [{self.status}]"
+
+    @property
+    @rounded
+    def pretax_price(self):
+        return (
+            convert_aftertax_to_pretax(self.price, self.tax_percentage)
+            if self.price and self.tax_percentage
+            else None
+        )
+
+    @property
+    @rounded
+    def total_price(self):
+        return sum([ol.price for ol in self.order_lines.all()], self.price)
+
+    @property
+    @rounded
+    def total_pretax_price(self):
+        return sum(
+            [ol.pretax_price for ol in self.order_lines.all()], self.pretax_price
+        )
+
+    @property
+    @rounded
+    def total_tax_percentage(self):
+        return ((self.total_price - self.total_pretax_price) / self.total_price) * 100
+
+    def clean(self):  # noqa: C901  is too complex
```

Comment:
is this now really too complex for flake? 😭  maybe enscapsulate it into some `_validate_order_lease` etc.?

Reply:
Yes:( I'll try to encapsulate it a bit.